docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+
+```typescript
+nftMetadataResults.filter(
+    (result): result is PromiseFulfilledResult<NftUpdate> =>
+      result.status === 'fulfilled',
+  )
+  .forEach((elm) =>
+    this.updateNft(
+      elm.value.nft,
+      ...
+    ),
+  );
+```
+
+#### Acceptable usages of `as`
+
+Although `as` is dangerous and discouraged from being used, the following is not intended as an exhaustive list of its valid use cases.
+
+##### `as` is acceptable to use for preventing or fixing `any` usage
+
+Type assertions are unsafe, but they are still always preferred to introducing `any` into the code.
+
+- With type assertions, we still get working intellisense, autocomplete, and other IDE and compiler features using the asserted type.
+- Type assertions also provide an indication of what the author intends or expects the type to be.
+- Often, the compiler will tell us exactly what the target type for an assertion needs to be, enabling us to avoid `as any`.
+
+```typescript
+// Error: Argument of type '"getNftInformation"' is not assignable to parameter of type 'keyof NftController'.ts(2345)
+// 'getNftInformation' is a private method of class 'NftController'
+sinon.stub(nftController, 'getNftInformation');
+```
+
+🚫 `as any`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as any);
+```
+
+✅ Compiler specifies that the target type should be `keyof NftController`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as keyof typeof nftController);
+```
+
+- Even an assertion to a wrong type still allows the compiler to show us warnings and errors as the code changes, and is therefore preferrable to the dangerous radio silence enforced by `any`.
+- For type assertions to an incompatible shape, use `as unknown as` as a last resort rather than `any` or `as any`.
+
+##### `as` is always acceptable to use for TypeScript syntax other than type assertion
+
+- Writing type guards often requires using the `as` keyword.
+
+```typescript
+function isFish(pet: Fish | Bird): pet is Fish {
+  return (pet as Fish).swim !== undefined;
+}
+```
+
+- Key remapping in mapped types uses the `as` keyword.
+
+```typescript
+type MappedTypeWithNewProperties<Type> = {
+  [Properties in keyof Type as NewKeyType]: Type[Properties];
+};
+```
+
+##### `as` is acceptable to use for typing data objects whose shape and contents are determined at runtime, externally, or through deserialization
+
+Preferably, this typing should be accompanied by runtime schema validation performed with type guards and unit tests.
+
+- e.g. The output of `JSON.parse()` or `await response.json()` for a known JSON input.
+- e.g. The type of a JSON file.
+
+<!-- TODO: Add example -->
+
+##### `as` may be acceptable to use in tests, for mocking or to exclude irrelevant but required properties from an input object
+
+It's recommended to provide accurate typing if there's any chance that omitting properties affects the accuracy of the test.
```

Comment:
Nit: This is a strange thing to say. Of course providing an invalid type would affect the accuracy of the test, how could it not? And how are we expecting people to make this determination?


When it comes to using `as` in tests, I had expected the guideline to be something like "`as` may be used to intentionally pass in invalid types in tests"

Reply:
I know that this guideline came out of a discussion that you and I had but I'm not sure I agree with it. If we are creating an object in a test and we need to satisfy a type, I feel that we should prefer a full representation of that type rather than a partial representation. Going further, though, if this results in tests that are difficult to maintain or hard to read, then this is a smell, and we should see if we can have the thing we are testing take a smaller portion of that object rather than the whole thing. I think that is a better route than "lying" in tests. Thoughts?